### DISCUSSION TITLE
How to enable bevy_dev_tools?

### DISCUSSION THREAD

**Discussion by fhgaha** [bevyengine/bevy#16053] [Q&A] [answered]
This [example](https://bevyengine.org/examples/ui-user-interface/ui/) says `Try enabling bevy_dev_tools`. How to do that?

**Comment by rparrett** [comment]
`bevy_dev_tools` is a [`cargo feature`](https://doc.rust-lang.org/cargo/reference/features.html).

You can enable it in your own projects by adding the feature to the Bevy dependency declaration in your `Cargo.toml` file.
```toml
bevy = { version = "0.14", features = ["bevy_dev_tools"] }
```

Or you can enable it when running examples in the Bevy repo with
```bash
cargo run --example ui --features=bevy_dev_tools
```

**Reply by fhgaha** [reply]
I have it there like that `bevy = { version = "0.14.2", features = ["dynamic_linking", "bevy_dev_tools"] }` but rust analyzer and the running example still saying "Try enabling feature bevy_dev_tools".

**Reply by BD103** [reply]
Could you paste a copy of your `Cargo.toml`?

**Reply by fhgaha** [reply]
Sure! 

```[package]
name = "water_slides"
version = "0.1.0"
edition = "2021"

[dependencies]
bevy = { version = "0.14.2", default-features = false,  features = [
	"dynamic_linking", 
	"bevy_dev_tools",
	
	#default
	"animation",
	"bevy_asset",
	"bevy_state",
	"bevy_audio",
	"bevy_color",
	"bevy_gilrs",
	"bevy_scene",
	"bevy_winit",
	"bevy_core_pipeline",
	"bevy_pbr",
	"bevy_gltf",
	"bevy_render",
	"bevy_sprite",
	"bevy_text",
	"bevy_ui",
	"multi_threaded",
	"png",
	"hdr",
	"vorbis",
	"x11",
	"bevy_gizmos",
	"android_shared_stdcxx",
	"tonemapping_luts",
	"smaa_luts",
	"default_font",
	"webgl2",
	"sysinfo_plugin",
] }
bevy_mod_raycast = "0.18.0"
bevy_lunex = { version = "0.2.4" }
bevy_rts_camera = { path = "crates/bevy_rts_camera" }
bevy_panorbit_camera = {path = "crates/bevy_panorbit_camera"}
bevy_dev_tools = "0.14.2"
bevy_egui = "0.30.0"

# Enable a small amount of optimization in the dev profile.
[profile.dev]
opt-level = 1

# Enable a large amount of optimization in the dev profile for dependencies.
[profile.dev.package."*"]
opt-level = 3
```

**Reply by BD103** [reply]
Does setting `default-features = true` remove the error? If so, then there's a chance your `Cargo.toml` is missing a default feature.

**Reply by fhgaha** [reply]
No, `default-features = true` does not fix that.

**Reply by BD103** [reply]
Very strange... are you importing from `bevy::dev_tools`, not `bevy_dev_tools`? What code are you running (a minimum reproducible example would be awesome!), and what exactly is the error message? I'm not sure I can diagnose without more information.

**Reply by fhgaha** [reply]
I'm running this [ui example](https://bevyengine.org/examples/ui-user-interface/ui/). It uses `#[cfg(feature = "bevy_dev_tools")]`. It doesn't do anything like `use bevy_dev_tools::*`. It's not exactly the error message, its text from the code to display, line 107: "Try enabling feature \"bevy_dev_tools\"."

I added 
```
use bevy_dev_tools::*;
use bevy::dev_tools::*;
```
but it doesn't seem to help

**Reply by rparrett** [reply]
The conditional compilation in that example won't work if you're running it outside of the Bevy project. You can either remove it:

```diff
- #[cfg(feature = "bevy_dev_tools")]
```

Or add a cargo feature to your own project that enables the bevy feature.

```toml
[features]
bevy_dev_tools = ["bevy/bevy_dev_tools"]
```
```bash
cargo run --features=bevy_dev_tools
```

**Reply by fhgaha** [reply]
By saying "running outside my bevy project" you mean I can't run example like `cargo run --example ui`? This is outside my bevy project? `examples` folder is right next to `src` folder and `Cargo.toml` file. Is this bevy reason or rust reason?

Also this 
```
[features]
bevy_dev_tools = ["bevy/bevy_dev_tools"]
```
didn't work:
```
error: failed to parse manifest at `D:\MyProjects\Bevy\my_project\Cargo.toml`

Caused by:
  failed to parse the version requirement `bevy/bevy_dev_tools` for dependency `bevy_dev_tools`

Caused by:
  unexpected character 'b' while parsing major version number
```

**Reply by BD103** [reply, marked as answer]
> By saying "running outside my bevy project" you mean I can't run example like `cargo run --example ui`? This is outside my bevy project?

I believe he means outside of the engine source code.

> ```
> Caused by:
> unexpected character 'b' while parsing major version number
> ```

It looks like you put `bevy_dev_tools = ["bevy/bevy_dev_tools"]` under `[dependencies]` instead of `[features]`.

**Reply by rparrett** [reply]
> Is this bevy reason or rust reason?

This is a rust thing. You can't do conditional compilation based on features of a dependency. So it works for examples within the Bevy repository/project, but not in a separate project with a Bevy dependency.

**Reply by fhgaha** [reply]
Understood. Thank you guys.
